tron/core/action.py

```diff
@@ -62,43 +62,52 @@ def copy(self):
     @staticmethod
     def from_json(state_data: str):
         """Deserialize a JSON string to an ActionCommandConfig instance."""
-        # Here we will also need to load the json for "command_config" since it's a string
         try:
             json_data = json.loads(state_data)
-            for k in json_data:
-                if k == "constraints":
-                    # convert back the list of dictionaries to a list of ConfigConstraint
-                    json_data[k] = [ConfigConstraint(**val) for val in json_data[k]]
-                elif k == "docker_parameters":
-                    # convert back the list of dictionaries to a list of ConfigParameter
-                    json_data[k] = [ConfigParameter(**val) for val in json_data[k]]
-                elif k == "extra_volumes":
-                    json_data[k] = [ConfigVolume(**val) for val in json_data[k]]
-                elif k == "secret_volumes":
-                    json_data[k] = [ConfigSecretVolume(**val) for val in json_data[k]]
-                elif k == "projected_sa_volumes":
-                    json_data[k] = [ConfigProjectedSAVolume(**val) for val in json_data[k]]
-                elif k == "node_affinities":
-                    json_data[k] = [ConfigNodeAffinity(**val) for val in json_data[k]]
-                elif k == "topology_spread_constraints":
-                    json_data[k] = [ConfigTopologySpreadConstraints(**val) for val in json_data[k]]
-                elif k == "secret_env":
-                    json_data[k] = {key: ConfigSecretSource(**val) for key, val in json_data[k].items()}
-                elif k == "field_selector_env":
-                    json_data[k] = {key: ConfigFieldSelectorSource(**val) for key, val in json_data[k].items()}
+            deserialized_data = {
+                "constraints": [
+                    ConfigConstraint(**val) for val in json_data["constraints"]
+                ],  # convert back the list of dictionaries to a list of ConfigConstraint
+                "docker_parameters": [ConfigParameter(**val) for val in json_data["docker_parameters"]],
+                "extra_volumes": [ConfigVolume(**val) for val in json_data["extra_volumes"]],
+                "node_affinities": [ConfigNodeAffinity(**val) for val in json_data["node_affinities"]],
+                "topology_spread_constraints": [
+                    ConfigTopologySpreadConstraints(**val) for val in json_data["topology_spread_constraints"]
+                ],
+                "secret_volumes": [ConfigSecretVolume(**val) for val in json_data["secret_volumes"]],
+                "projected_sa_volumes": [ConfigProjectedSAVolume(**val) for val in json_data["projected_sa_volumes"]],
+                "secret_env": {key: ConfigSecretSource(**val) for key, val in json_data["secret_env"].items()},
+                "field_selector_env": {
+                    key: ConfigFieldSelectorSource(**val) for key, val in json_data["field_selector_env"].items()
```

Comment:
on another look, i realized there's another potential hazard here: if we're rolling back from a tron that adds new fields to any of these, i think that `CONSTRUCTOR(**vals)` will throw an exception due to the extra keys, right?

I have couple ideas on how to handle this if so (e.g., new classmethods on these that take a dict and construct the object, classmethod that returns the currently supported keys so that we can filter unsupported keys out, etc) - but not particularly tied to any option here :)